spi/imx6ull-qspi/qspi.c

```diff
@@ -0,0 +1,363 @@
+	{ { pctl_pad_nand_re, 1, 4, 0 }, { pctl_pad_nand_we, 1, 4, 0 }, { pctl_pad_nand_d0, 1, 4, 0 }, { pctl_pad_nand_d1, 1, 4, 0 }, { pctl_pad_nand_d2, 1, 4, 0 }, { pctl_pad_nand_d3, 1, 4, 0 }, { pctl_pad_nand_d4, 1, 4, 0 }, { pctl_pad_nand_d5, 1, 4, 0 } },
+};
+
+
+struct {
+	volatile uint32_t *base;
+	uint32_t flash_base_addr[2];
+	int init;
+} qspi_common;
+
+/* TODO set ohm value in the flash register */
+/* TODO find clock frequency - ~50Mhz */
+
+
+static int wait_cmd_done_busy(void)
+{
+	int max_iter;
+
+	for (max_iter = 1000; max_iter > 0; max_iter--) {
+		if ((*(qspi_common.base + QSPI_SR) & QSPI_SR_BUSY) == 0) {
+			return EOK;
+		}
+		usleep(100);
+	}
+	printf("WAIT BUSY TIMEOUT\n");
+	return -ETIME;
+}
+
+
+static void set_mux(qspi_dev_t dev_no)
+{
+	platformctl_t ctl;
+	int i;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_iomux;
+
+	for (i = 0; i < 8; i++) {
+		ctl.iomux.mux = qspi_pctl_mux[dev_no][i].mux;
+		ctl.iomux.mode = qspi_pctl_mux[dev_no][i].val;
+		ctl.iomux.sion = qspi_pctl_mux[dev_no][i].sion;
+		platformctl(&ctl);
+	}
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_iopad;
+
+	for (i = 0; i < 8; i++) {
+		ctl.iopad.pad = qspi_pctl_pad[dev_no][i].pad;
+		ctl.iopad.speed = qspi_pctl_pad[dev_no][i].speed;
+		ctl.iopad.dse = qspi_pctl_pad[dev_no][i].dse;
+		ctl.iopad.sre = qspi_pctl_pad[dev_no][i].sre;
+		platformctl(&ctl);
+	}
+}
+
+static void set_clk(void)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_devclock;
+	ctl.devclock.dev = pctl_clk_qspi;
+	ctl.devclock.state = 0x03;
+
+	platformctl(&ctl);
+}
+
+
+int qspi_setLutSeq(const lut_seq_t *lut, unsigned int lut_seq)
+{
+	int i;
+
+	if (lut_seq >= NUM_LUT_SEQ) {
+		return -EINVAL;
+	}
+	if (*(qspi_common.base + QSPI_LCKCR) & QSPI_LCKCR_LOCK) {
+		*(qspi_common.base + QSPI_LUTKEY) = QSPI_LUTKEY_KEY;
+		*(qspi_common.base + QSPI_LCKCR) = (*(qspi_common.base + QSPI_LCKCR) & ~(QSPI_LCKCR_UNLOCK | QSPI_LCKCR_LOCK)) | QSPI_LCKCR_UNLOCK;
+	}
+
+	for (i = 0; i < 4; i++) {
+		*(qspi_common.base + QSPI_LUTn(QSPI_SEQ_START_REGN(lut_seq) + i)) = (((uint32_t)lut->instrs[2 * i + 1]) << 16) | lut->instrs[2 * i];
+	}
+
+	*(qspi_common.base + QSPI_LUTKEY) = QSPI_LUTKEY_KEY;
+	*(qspi_common.base + QSPI_LCKCR) = (*(qspi_common.base + QSPI_LCKCR) & ~(QSPI_LCKCR_UNLOCK | QSPI_LCKCR_LOCK)) | QSPI_LCKCR_LOCK;
+
+	return EOK;
+}
+
+
+static void set_IPCR(int seq_num, size_t idatsz)
+{
+	uint32_t reg = *(qspi_common.base + QSPI_IPCR);
+
+	reg = (reg & ~(0x0f << 24)) | ((seq_num & 0x0f) << 24);
+	reg = (reg & ~(0xffff)) | (idatsz & 0xffff);
+	*(qspi_common.base + QSPI_IPCR) = reg;
+}
+
+
+void qspi_setTCSH(uint8_t cycles)
+{
+	*(qspi_common.base + QSPI_FLSHCR) = (*(qspi_common.base + QSPI_FLSHCR) & ~(0x0f << 8)) | (cycles << 8);
+}
+
+
+void qspi_setTCSS(uint8_t cycles)
+{
+	*(qspi_common.base + QSPI_FLSHCR) = (*(qspi_common.base + QSPI_FLSHCR) & ~(0x0f)) | cycles;
+}
+
+
+int _qspi_readBusy(qspi_dev_t dev, unsigned int lut_seq, uint32_t addr, void *buf, size_t size)
```

Comment:
> Hmm, why is this function called `readBusy`? Is it because of temporary busy waiting at the end? If so:
> 
> * write function does not have `Busy` postfix,
> * there are plans to replace busy waiting with interrupts, so this name will become obsolete quick

Both checking if there is data to read and waiting for command to finish is done busy. I'd like to eventually add separate function _qspi_read as depending on the size of read one or other shall be faster. Eg for Get id command, get status register command or command that don't actually read anything from the device busy read shall be used.

Reply:
> Also, we use `_` prefix to indicate unlocked functions - this lib doesn't have any locks, so either all functions should be marked as unlocked or none (I prefer the latter option)

Can you elaborate why you prefer marking none?

Reply:
> > Hmm, why is this function called `readBusy`? Is it because of temporary busy waiting at the end? If so:
> > 
> > * write function does not have `Busy` postfix,
> > * there are plans to replace busy waiting with interrupts, so this name will become obsolete quick
> 
> Both checking if there is data to read and waiting for command to finish is done busy. I'd like to eventually add separate function _qspi_read as depending on the size of read one or other shall be faster. Eg for Get id command, get status register command or command that don't actually read anything from the device busy read shall be used.

However I think that non busy read shall be added in future iteration.

Reply:
About marking unlocked functions: because it doesn't really make sense to mark functions as unlocked if whole lib isn't thread-safe.

Examples from existing code base (I'm refraining from providing my own code as example):
https://github.com/phoenix-rtos/phoenix-rtos-devices/blob/master/adc/ad7779/imx6ull/sai.c
https://github.com/phoenix-rtos/phoenix-rtos-devices/blob/master/storage/imxrt-flash/fspi.c
+ pretty much all single threaded servers and their helper functions.

I also found a counter-example:
https://github.com/phoenix-rtos/phoenix-rtos-devices/blob/master/tty/pc-tty/ttypc_vtf.c

I agree that it's not 100% clarified in our guidelines what should be done in such situation, but it has been agreed to mark _unlocked_ versions of functions. I wouldn't really call a function unlocked if there's no locks in sight.

Also, we violate C rule of `_` prefix being reserved (along with `_t` suffix we use), so this convention will need to change at some point.